src/modules/dwm.cpp

```diff
@@ -0,0 +1,544 @@
+      m_log.info("%s: Building tags label", name());
+      int cmd_count = 0;
+      if (m_tags_scroll) {
+        bool ignore_empty_tags = !m_tags_scroll_empty && m_conf.get<string>(name(), "label-empty", "").empty();
+        const auto* next =
+            m_tags_scroll_reverse ? prev_scrollable_tag(ignore_empty_tags) : next_scrollable_tag(ignore_empty_tags);
+        const auto* prev =
+            m_tags_scroll_reverse ? next_scrollable_tag(ignore_empty_tags) : prev_scrollable_tag(ignore_empty_tags);
+        if (next != nullptr) {
+          ++cmd_count;
+          builder->action(mousebtn::SCROLL_DOWN, *this, EVENT_TAG_VIEW, to_string(next->bit_mask));
+        }
+        if (prev != nullptr) {
+          ++cmd_count;
+          builder->action(mousebtn::SCROLL_UP, *this, EVENT_TAG_VIEW, to_string(prev->bit_mask));
+        }
+      }
+
+      bool first = true;
+      for (const auto& tag : m_tags) {
+        // Don't insert separator before first tag
+        if (first) {
+          first = false;
+        } else if (*m_seperator_label) {
+          builder->node(m_seperator_label);
+        }
+
+        if (m_tags_click) {
+          builder->action(mousebtn::LEFT, *this, EVENT_TAG_VIEW, to_string(tag.bit_mask));
+          builder->action(mousebtn::RIGHT, *this, EVENT_TAG_TOGGLE_VIEW, to_string(tag.bit_mask));
+          builder->node(tag.label);
+          builder->action_close();
+          builder->action_close();
+        } else {
+          builder->node(tag.label);
+        }
+      }
+      for (; cmd_count; --cmd_count) {
+        builder->action_close();
+      }
+    } else {
+      return false;
+    }
+    return true;
+  }
+
+  bool dwm_module::input(const string& action, const string& data) {
+    m_log.info("%s: Sending workspace %s command to ipc handler", name(), action);
+
+    try {
+      m_ipc->run_command(action, (Json::UInt64)stoul(data));
+      return true;
+    } catch (const dwmipc::SocketClosedError& err) {
+      m_log.err("%s: Disconnected from socket: %s", name(), err.what());
+      sleep(chrono::duration<double>(1));
+      reconnect_dwm();
+    } catch (const dwmipc::IPCError& err) {
+      throw module_error(err.what());
+    }
+
+    return false;
+  }
+
+  dwm_module::state_t dwm_module::get_state(tag_mask_t bit_mask) const {
+    /**
+     * ---------------------------------------------------------------
+     * |             Tag              |                  |           |
+     * |------------------------------| Monitor Selected |   State   |
+     * | Urgent | Selected | Occupied |                  |           |
+     * |--------|----------|----------|------------------|-----------|
+     * |   Yes  |     *    |     *    |         *        |   Urgent  |
+     * |   No   |    Yes   |     *    |        Yes       |  Focused  |
+     * |   No   |    Yes   |     *    |        No        | Unfocused |
+     * |   No   |    No    |    Yes   |         *        |  Visible  |
```

Comment:
Is this dwm terminology? From what I can tell tags in this state don't necessarily have to be visible (or am I understanding dwm's tag system wrong?).